script.js

```diff
@@ -1,4 +1,577 @@
+  }
+  return rollStateMessage;
+};
+
+const autoGenerateOptimalCombinedNumber = function () {
+  // MUST initialize this as an empty string or else the value will be undefined during reassignment later
+  let currentPlayerFinalNum = "";
+
+  // Arrange digits from big to small in the array, convert each of them to string and keep them in currentPlayerFinalNum variable
+  if (gameMode === HIGHEST_COMBINED_NUMBER_MODE) {
+    currentPlayerDiceNumbers.sort((a, b) => b - a);
+    for (i = 0; i < currentPlayerDiceNumbers.length; i += 1) {
+      currentPlayerFinalNum += String(currentPlayerDiceNumbers[i]);
+    }
+  }
+  // Arrange digits from small to big in the array, etc..
+  else if (gameMode === LOWEST_COMBINED_NUMBER_MODE) {
+    currentPlayerDiceNumbers.sort((a, b) => a - b);
+    for (i = 0; i < currentPlayerDiceNumbers.length; i += 1) {
+      currentPlayerFinalNum += String(currentPlayerDiceNumbers[i]);
+    }
+  }
+
+  // Transfer the final num to another variable for each player separately for LEADERBOARD later
+  if (currentPlayer === 1) {
+    player1FinalNumEachRound.push(Number(currentPlayerFinalNum));
+  } else if (currentPlayer === 2) {
+    player2FinalNumEachRound.push(Number(currentPlayerFinalNum));
+  }
+  return `Player ${currentPlayer}, your number is ${currentPlayerFinalNum}. <br>
+          Player 2's turn now! <br>
+          Please hit the "Submit" button to continue.`;
+};
+
+const showLeaderboard = function () {
+  // Calculate the sum of all player 1 optimal combined numbers
+  let player1Sum = 0;
+  for (i = 0; i < player1FinalNumEachRound.length; i += 1) {
+    player1Sum += player1FinalNumEachRound[i];
+  }
+
+  // Calculate the sum of all player 2 optimal combined numbers
+  let player2Sum = 0;
+  for (i = 0; i < player2FinalNumEachRound.length; i += 1) {
+    player2Sum += player2FinalNumEachRound[i];
+  }
+
+  let leaderboardMessage = `Player 1's score: ${player1Sum} | Player 2's score: ${player2Sum} <br>
+          Press Submit to play again. <br> <br>
+          <b> ------ LEADERBOARD ------ </b> <br>`;
+
+  // Player 1 is leading
+  if (
+    (gameMode === HIGHEST_COMBINED_NUMBER_MODE && player1Sum > player2Sum) ||
+    (gameMode === LOWEST_COMBINED_NUMBER_MODE && player1Sum < player2Sum)
+  ) {
+    leaderboardMessage += `Player 1 is currently winning!!! <br>
+          Player 1: ${player1Sum} <br>
+          Player 2: ${player2Sum}`;
+  }
+
+  // Player 2 is leading
+  else if (
+    (gameMode === HIGHEST_COMBINED_NUMBER_MODE && player1Sum < player2Sum) ||
+    (gameMode === LOWEST_COMBINED_NUMBER_MODE && player1Sum > player2Sum)
+  ) {
+    leaderboardMessage += `Player 2 is currently winning!!! <br>
+          Player 2: ${player2Sum} <br>
+          Player 1: ${player1Sum}`;
+  }
+
+  // Draw condition for any mode
+  else if (player1Sum === player2Sum) {
+    leaderboardMessage += `DRAW! No one is currently winning!!!`;
+  }
+  return leaderboardMessage;
+};
```

Comment:
This can be its own function `determineLeader(player1Sum, player2Sum)` or something like that.